pkg/issuer/acme/dns/livedns/livedns.go

```diff
@@ -0,0 +1,203 @@
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+)
+
+// Gandi API reference:       http://doc.livedns.gandi.net/
+
+var (
+	// endpoint is the Gandi API endpoint used by Present and
+	// CleanUp. It is overridden during tests.
+	endpoint = "https://dns.api.gandi.net/api/v5"
+
+	// findZoneByFqdn determines the DNS zone of an fqdn. It is overridden
+	// during tests.
+	findZoneByFqdn = util.FindZoneByFqdn
+)
+
+// inProgressInfo contains information about an in-progress challenge
+type inProgressInfo struct {
+	fieldName string
+	authZone  string
+}
+
+// DNSProvider is an implementation of the
+// acme.ChallengeProviderTimeout interface that uses Gandi's LiveDNS
+// API to manage TXT records for a domain.
+type DNSProvider struct {
+	apiKey          string
+	inProgressFQDNs map[string]inProgressInfo
+	inProgressMu    sync.Mutex
+	client          *http.Client
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for Gandi.
+func NewDNSProviderCredentials(apiKey string) (*DNSProvider, error) {
+	if apiKey == "" {
+		return nil, fmt.Errorf("Gandi DNS: No Gandi API Key given")
+	}
+	return &DNSProvider{
+		apiKey:          apiKey,
+		inProgressFQDNs: make(map[string]inProgressInfo),
+		client:          &http.Client{Timeout: 10 * time.Second},
+	}, nil
+}
+
+// Present creates a TXT record using the specified parameters.
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl := util.DNS01Record(domain, keyAuth)
+	if ttl < 300 {
+		ttl = 300 // 300 is gandi minimum value for ttl
+	}
+
+	// find authZone
+	authZone, err := findZoneByFqdn(fqdn, util.RecursiveNameservers)
+	if err != nil {
+		return fmt.Errorf("Gandi DNS: findZoneByFqdn failure: %v", err)
+	}
+
+	// determine name of TXT record
+	if !strings.HasSuffix(
+		strings.ToLower(fqdn), strings.ToLower("."+authZone)) {
+		return fmt.Errorf(
+			"Gandi DNS: unexpected authZone %s for fqdn %s", authZone, fqdn)
+	}
+	name := fqdn[:len(fqdn)-len("."+authZone)]
+
+	// acquire lock and check there is not a challenge already in
+	// progress for this value of authZone
+	d.inProgressMu.Lock()
+	defer d.inProgressMu.Unlock()
+
+	// add TXT record into authZone
+	err = d.addTXTRecord(util.UnFqdn(authZone), name, value, ttl)
+	if err != nil {
+		return err
+	}
+
+	// save data necessary for CleanUp
+	d.inProgressFQDNs[fqdn] = inProgressInfo{
+		authZone:  authZone,
+		fieldName: name,
+	}
+	return nil
+}
+
+// CleanUp removes the TXT record matching the specified parameters.
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, _ := util.DNS01Record(domain, keyAuth)
+
+	// acquire lock and retrieve authZone
+	d.inProgressMu.Lock()
+	defer d.inProgressMu.Unlock()
+	if _, ok := d.inProgressFQDNs[fqdn]; !ok {
+		// if there is no cleanup information then just return
+		return nil
```

Comment:
If cert-manager crashes and restarts, it will not clean up challenges that should be cleaned up because the `inProgressFQDN` map is stored in memory 😬

Reply:
:+1: removed this